### DISCUSSION TITLE
Questions regarding time

### DISCUSSION THREAD

**Discussion by sagatowski** [tcunit/TcUnit#192] [Q&A] [answered]
Question asked by e-mail:
Helo Jakob,

I am currently writing my diploma thesis and I am using the TcUnit application you developed. I really like what you have created, I will make great use of it!
However, I have not yet found answers to some questions.:
1:
For example, I have a function block (FB) which, after x time, sets one of its outputs to TRUE in response. 
I would like to measure this "x" time to see if it is accurate. So I would wait for that "x" time and then check the output.

2:
Also I use the "TC Measurement" application to visualize/export the test results. For this I would need some delay between TESTs. So when a test is run it should wait e.g. 10ms, then start the next test (I found a partial solution to this problem, RUN_IN_SEQUENCE(), but then each test should be in a separate FB? I currently use Method, and it would be nice if it could stay that way. ) Thanks for your help and reply.

**Comment by sagatowski** [comment, marked as answer]
For 1:
Simply create a `TON `and once the `TON `has triggered, simply evaluate (assert) the output.
So simply something like

```
fbTon : TON := (PT := T#10S);
fbTestBlock : FB_YourBlockUnderTest;
---
TEST('HelloWorld');
fbTon(IN := TRUE);
IF fbTon.Q THEN
    AssertEquals(Expected := TRUE,
                         Actual := fbTestBlock.bBool,
                         Message := 'Not Matching');
END_IF
TEST_FINISHED();
```


For 2: The only two ways are `RUN_IN_SEQUENCE()` or by using [this way](https://tcunit.org/frequently-asked-questions/?Display_FAQ=1008).
You could also run `TEST_ORDERED()` and simply add a TON at the end of the test to delay the output of the test. Please [see this](https://tcunit.org/frequently-asked-questions/?Display_FAQ=970).

**Reply by Pippen19** [reply]
Should it not be done this way:
```
fbTon : TON := (PT := T#10S);
fbTestBlock : FB_YourBlockUnderTest;
---
TEST('HelloWorld');
fbTon(IN := TRUE);
IF fbTon.Q THEN
    AssertEquals(Expected := TRUE,
                         Actual := fbTestBlock.bBool,
                         Message := 'Not Matching');
    TEST_FINISHED();
END_IF
```
"TEST_FINISHED();" should be in IF statement, so that the test does not end after the first cycle, right?

**Reply by sagatowski** [reply]
> Should it not be done this way:
> 
> ```
> fbTon : TON := (PT := T#10S);
> fbTestBlock : FB_YourBlockUnderTest;
> ---
> TEST('HelloWorld');
> fbTon(IN := TRUE);
> IF fbTon.Q THEN
>     AssertEquals(Expected := TRUE,
>                          Actual := fbTestBlock.bBool,
>                          Message := 'Not Matching');
>     TEST_FINISHED();
> END_IF
> ```
> 
> "TEST_FINISHED();" should be in IF statement, so that the test does not end after the first cycle, right?

You're correct!

**Comment by bumbasor** [comment]
Hi, I have a similar problem with timer and am also working on my master thesis :)

I have a function block that has a part with TON timer and after timer is finished, output variable receives some value.
But, when I try to add TON to unit test as you showed, my asserts under 'IF fbTon.Q ' are not checked because time of tests being run is much lower than my time constant for timer. Is there any way to check that output value that should be written in the variable with unit test?

**Reply by I-Campbell** [reply]
I think TEST_FINISHED(); should go inside the IF statement after the Assert call.

**Reply by bumbasor** [reply]
I tried that also previously, but then maximum number of asserts is exceeded.

**Reply by sagatowski** [reply]
@bumbasor Check this: https://tcunit.org/#/faq?id=_6-when-i-run-more-than-100-tests-in-a-single-test-suite-i-get-the-wrong-results-why

**Reply by bumbasor** [reply]
Does it make sense to copy function block and use a 'for' loop instead of timer and use that FB for testing? Or is that not the proper way to test it? Thanks :)

**Reply by bumbasor** [reply]
I increased max number of asserts to 60 000 and put TEST_FINISHED(); inside IF tTon.Q after Assert call. Now it is running for a few minutes and again the same error appears. Looks like the test is not finishing at all. PT of my timer is set to 10ms. I guess I am doing something wrong. Any help would be appreciated :)
And also, is it an ok method simulating timer with for loop?

**Reply by sagatowski** [reply]
Are you sure you are running the timer? Do you have the timer declared as an instance variable?

**Reply by bumbasor** [reply]
I basically did everything like you showed on this example above.  The only difference is that I put TEST_FINISHED(); inside IF statement because otherwise it will not check asserts inside IF statement.

**Reply by sagatowski** [reply]
Please answer this question:
Do you have the timer declared as an instance variable?

**Reply by bumbasor** [reply]
Yes, I have. 
tTon	 :TON := (PT := T#10MS);

**Reply by sagatowski** [reply]
That doesn't answer the question.
Is it declared as an instance variable or as a local method variable?

**Reply by bumbasor** [reply]
Now I get it. Previously I declared it in the method of a FB.
Thanks a lot! :)

**Reply by sagatowski** [reply]
You're welcome.
